### DISCUSSION TITLE
Is it possible to have !Send handler for requests ?

### DISCUSSION THREAD

**Discussion by artem-tim** [tokio-rs/axum#2522] [Q&A] [answered]
### Summary

I use a http request processing library which has a !Send constraint. Is it possible to run a multi-threaded axum server and have it assign a dedicated thread for each incoming request that would explicitly allow !Send ? I've initially created a Service, but then noticed that Router Layer cannot be used because of Send constraint for the service : `L::Service: Service<Request> + Clone + Send + 'static`

### axum version

0.7

**Comment by davidpdrsn** [comment, marked as answer]
No that’s not possible. axum requires `Send`. Allowing `!Send` would basically require duplicating all the code because of how auto traits work in trait objects.

I think your best option is to spawn a dedicated thread and communicate with it via channels.

**Reply by artem-tim** [reply]
Thanks. In that case, would you recommend maybe using hyper directly rather than axum ?

**Reply by davidpdrsn** [reply]
Depends on your use case. Hyper is very low level.

**Reply by artem-tim** [reply]
Hmm my usecase is basically just that : apply a !Send http request processing function which returns an http::Response. The library was originally designed to work in web-worker context as wasm (which is inherently single threaded), and I'm trying emulate this single-threaded runtime on a custom http server. Still unsure of the best libraries for the job, so I wonder if axum is an appropriate solution
